### ISSUE TITLE
Does batch size have a huge impact for result？

### ISSUE THREAD

**Issue by theSha1do1w** [jiawei-ren/BalancedMSE#5]
We can see that you use a classifying within a batch in the paper, so does batch size have a huge impact for result？?

**Comment by jiawei-ren** [comment]
The batch size's impact is small in practice. Our BMC loss even works in a Triplet Loss way, where there are only two samples: a positive anchor and a negative anchor.

For verification, we fixed the training batch size and change the number of samples used in BMC. We show the 1-d linear regression results in the following table. The performance is stable as long as a positive anchor, i.e., the ground truth y_target is sampled.

<img width="551" alt="Screenshot 2022-05-02 at 11 04 28 AM" src="https://user-images.githubusercontent.com/72253125/166179596-d021e5bf-498c-4090-8442-708c076bfe19.png">

**Comment by jiawei-ren** [comment]
Therefore, when the batch size **B** or the regression target's dimension **D** is too large and computing a full **O(DxB^2)** similarity matrix is not affordable, we can safely downsample the batch for BMC. The down-sampled version BMC:

```python
def bmc_loss_downsample(pred, target, noise_var, num_sample=2):
    samples = target[:num_sample]   # only use the first num_sample elements as samples
    samples = samples.T.repeat(pred.shape[0], 1)   # expand to a sample matrix
    samples[num_sample:, :1] = target[num_sample:]    # add positive samples to unsampled elements

    mse_term = 0.5 * (pred - target).pow(2) / noise_var
    balancing_term = - 0.5 * (pred - samples).pow(2) / noise_var
    balancing_term = torch.logsumexp(balancing_term, dim=-1, keepdim=True)

    loss = mse_term + balancing_term
    loss = loss * (2 * noise_var).detach()

    return loss.mean()
```

The results on IMDB-WIKI-DIR:
|      | B=256 | B=128 | B=64  | B=32  | B=16  | B=8   | B=4   | B=2   |
|------|-------|-------|-------|-------|-------|-------|-------|-------|
| bMAE | 12.69 | 12.67 | 12.69 | 12.64 | 12.67 | 12.67 | 12.67 | 12.69 |

**Comment by Dong-Shawn** [comment]
Got it, thanks for your reply!!

**Comment by theSha1do1w** [comment]
Got it! Thx for your reply!!
